### PR TITLE
server: allow to override threads server pool with --threads-http

### PR DESCRIPTION
**Pull request by phymbert**
### Motivation

The default [httplib threads pool](https://github.com/yhirose/cpp-httplib/tree/master?tab=readme-ov-file#default-thread-pool-support) is set to `std::thread::hardware_concurrency()`
which is NOT always adapted for completions.
As far, completions tasks can be deferred, http server threads pool
can quickly be full by reaching the number of available host cpu cores
while actually continuous batching can accept more requests on the GPU device.

This fix allows the user to choose the balanced number of threads in the http server pool.

### Changes
Introduce `--threads-http N` to override the default configuration.

### CODE REVIEW

File: examples/server/server.cpp
Comment (ngxson) [inline examples/server/server.cpp:2069]: On the doc, it's `--threads-server` but here `--threads-http`

I'd prefer the name `--threads-http` though, because saying "server" is quite unclear (easy to be confused with inference threads)

File: examples/server/server.cpp
Comment (ngxson) [inline examples/server/server.cpp:3525]: Be default, I think number of http threads should be `n_parallel + 2` just to be sure, because the worst case:
- We have `n_parallel` requests running at the same time
- Plus one request for `/health`
- Plus one maybe for metrics endpoint

Comment (ngxson) [reply examples/server/server.cpp:3525]: P/s: These threads does almost nothing than just sitting and wait for `recv()`, so logically you can spawn how many you want without compromise performance.

In fact, modern async web framework doesn't even need thread pool, because it's done by using `epoll`. That should be ideal for use, but one major problem is that these framework are usually big, while what we want is just a single .hpp file

Comment (phymbert) [reply examples/server/server.cpp:3525]: Good point, but I prefer to keep the default httplib behavior for now and let the user decide if require. I would like to avoid people complaining to see lot of CPU threads created by the server because they are playing hard with GPU sequences and/or continuous batching.

**Review by ngxson — APPROVED** [review]

**Comment by phymbert** [comment]
@ggerganov Hi, as I got one approval, should I wait for yours ? thanks

**Comment by ggerganov** [comment]
If a PR touches just the `server` code, you don't need to wait for my approval.

Btw, isn't it better to reuse the `--parallel` parameter instead of introducing a new one?

**Comment by phymbert** [comment]
> If a PR touches just the `server` code, you don't need to wait for my approval.

Noted thanks.
 
> Btw, isn't it better to reuse the `--parallel` parameter instead of introducing a new one?

No there are different meaning, `--parallel` for number of sequences/slots on the llama backend, `--threads-http` for number of concurrent http requests on httplib (which can generate deferred tasks if all slots are processing).

**Pull request merged**
